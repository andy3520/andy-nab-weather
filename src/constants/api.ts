export const ROOT_URL = process.env.REACT_APP_ROOT_URL;

export enum EHTTPStatus {
  OK = 200,
  NOT_FOUND = 404,
  LIMIT_EXCEED = 104,
}

export enum EApiMessages {
  DEFAULT = "Couldn't fetch weather data!",
  NOT_FOUND = 'Location not found',
  LIMIT_EXCEED = 'Detect Location API exceed limit. Please use search.',
  CANNOT_DETECT_LOCATION = "Couldn't detect your location by ip address, try to search you location!",
}
