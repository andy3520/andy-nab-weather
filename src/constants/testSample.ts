import { IIPInfo, IWeatherLoc, IWeatherLocSearch } from '../types';

export const ipLocSample: IIPInfo = {
  ip: '113.185.78.133',
  type: 'ipv4',
  continent_code: 'AS',
  continent_name: 'Asia',
  country_code: 'VN',
  country_name: 'Vietnam',
  region_code: 'SG',
  region_name: 'Ho Chi Minh',
  city: 'Ho Chi Minh City',
  zip: null,
  latitude: 10.779999732971191,
  longitude: 106.65499877929688,
  location: {
    geoname_id: 1566083,
    capital: 'Hanoi',
    languages: [
      { code: 'vi', name: 'Vietnamese', native: 'Ti\u1ebfng Vi\u1ec7t' },
    ],
    country_flag: 'http://assets.ipstack.com/flags/vn.svg',
    country_flag_emoji: '\ud83c\uddfb\ud83c\uddf3',
    country_flag_emoji_unicode: 'U+1F1FB U+1F1F3',
    calling_code: '84',
    is_eu: false,
  },
};

export const locSample: IWeatherLocSearch[] = [
  {
    title: 'Ho Chi Minh City',
    location_type: 'City',
    woeid: 1252431,
    latt_long: '10.759180,106.662498',
  },
];

export const locWeatherSample: IWeatherLoc = {
  consolidated_weather: [
    {
      id: 5058293731950592,
      weather_state_name: 'Heavy Cloud',
      weather_state_abbr: 'hc',
      wind_direction_compass: 'NNE',
      created: '2021-01-01T13:00:46.267827Z',
      applicable_date: '2021-01-01',
      min_temp: 21.125,
      max_temp: 29.375,
      the_temp: 28.465,
      wind_speed: 5.7640538130014045,
      wind_direction: 12.15257635959627,
      air_pressure: 1012.5,
      humidity: 57,
      visibility: 11.298703074047562,
      predictability: 71,
    },
    {
      id: 6373796329029632,
      weather_state_name: 'Heavy Cloud',
      weather_state_abbr: 'hc',
      wind_direction_compass: 'NNE',
      created: '2021-01-01T13:00:49.267235Z',
      applicable_date: '2021-01-02',
      min_temp: 22.34,
      max_temp: 30.395000000000003,
      the_temp: 29.96,
      wind_speed: 4.656074916882738,
      wind_direction: 16.497329055585254,
      air_pressure: 1011.5,
      humidity: 55,
      visibility: 14.153282331185874,
      predictability: 71,
    },
    {
      id: 6529977378930688,
      weather_state_name: 'Showers',
      weather_state_abbr: 's',
      wind_direction_compass: 'NNE',
      created: '2021-01-01T13:00:52.654744Z',
      applicable_date: '2021-01-03',
      min_temp: 22.91,
      max_temp: 32.215,
      the_temp: 30.174999999999997,
      wind_speed: 3.195766099199343,
      wind_direction: 25.042992793059483,
      air_pressure: 1011.0,
      humidity: 61,
      visibility: 13.731371291656725,
      predictability: 73,
    },
    {
      id: 4865342410063872,
      weather_state_name: 'Light Rain',
      weather_state_abbr: 'lr',
      wind_direction_compass: 'NNE',
      created: '2021-01-01T13:00:55.308657Z',
      applicable_date: '2021-01-04',
      min_temp: 23.479999999999997,
      max_temp: 32.89,
      the_temp: 30.985,
      wind_speed: 4.010256309334818,
      wind_direction: 31.309153743191175,
      air_pressure: 1010.5,
      humidity: 61,
      visibility: 13.664573888491212,
      predictability: 75,
    },
    {
      id: 5948524706398208,
      weather_state_name: 'Heavy Rain',
      weather_state_abbr: 'hr',
      wind_direction_compass: 'NNE',
      created: '2021-01-01T13:00:58.682866Z',
      applicable_date: '2021-01-05',
      min_temp: 23.64,
      max_temp: 31.155,
      the_temp: 28.270000000000003,
      wind_speed: 4.794026365712241,
      wind_direction: 27.60297747685865,
      air_pressure: 1011.0,
      humidity: 74,
      visibility: 12.07790254911318,
      predictability: 77,
    },
    {
      id: 5815312269180928,
      weather_state_name: 'Heavy Rain',
      weather_state_abbr: 'hr',
      wind_direction_compass: 'ENE',
      created: '2021-01-01T13:01:01.655153Z',
      applicable_date: '2021-01-06',
      min_temp: 23.79,
      max_temp: 32.555,
      the_temp: 30.26,
      wind_speed: 2.0922822715342404,
      wind_direction: 59.5,
      air_pressure: 1011.0,
      humidity: 66,
      visibility: 9.999726596675416,
      predictability: 77,
    },
  ],
  time: '2021-01-01T22:03:34.235609+07:00',
  sun_rise: '2021-01-01T06:11:58.923405+07:00',
  sun_set: '2021-01-01T17:42:13.362750+07:00',
  timezone_name: 'LMT',
  parent: {
    title: 'Vietnam',
    location_type: 'Country',
    woeid: 23424984,
    latt_long: '15.974210,107.868042',
  },
  sources: [
    {
      title: 'BBC',
      slug: 'bbc',
      url: 'http://www.bbc.co.uk/weather/',
      crawl_rate: 360,
    },
    {
      title: 'Forecast.io',
      slug: 'forecast-io',
      url: 'http://forecast.io/',
      crawl_rate: 480,
    },
    {
      title: 'Met Office',
      slug: 'met-office',
      url: 'http://www.metoffice.gov.uk/',
      crawl_rate: 180,
    },
    {
      title: 'OpenWeatherMap',
      slug: 'openweathermap',
      url: 'http://openweathermap.org/',
      crawl_rate: 360,
    },
    {
      title: 'World Weather Online',
      slug: 'world-weather-online',
      url: 'http://www.worldweatheronline.com/',
      crawl_rate: 360,
    },
  ],
  title: 'Ho Chi Minh City',
  location_type: 'City',
  woeid: 1252431,
  latt_long: '10.759180,106.662498',
  timezone: 'Asia/Ho_Chi_Minh',
};
