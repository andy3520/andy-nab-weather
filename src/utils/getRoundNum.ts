export const getRoundUp = (num: number): number => {
  return !num ? 0 : Math.ceil(num);
};
