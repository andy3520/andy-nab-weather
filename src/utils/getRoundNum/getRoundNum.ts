export const getRoundNum = (num: number): number => {
  // eslint-disable-next-line no-restricted-globals
  if (typeof num !== 'number' || isNaN(num)) return 0;

  return Math.round(num);
};
