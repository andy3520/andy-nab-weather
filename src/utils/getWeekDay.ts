import { WEEK_DAYS } from '../constants/weekDay';

export const getWeekDay = (date: string): string => {
  if (!date) return '';

  const parsedDate = new Date(date);
  const weekDayIndex = parsedDate.getDay();

  return WEEK_DAYS[weekDayIndex];
};
