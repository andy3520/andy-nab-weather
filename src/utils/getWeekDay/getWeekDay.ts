import { WEEK_DAYS } from '../../constants/weekDay';

export const getWeekDay = (date: string): string => {
  if (!date || !Date.parse(date)) return '';

  const parsedDate = new Date(date);
  const weekDayIndex = parsedDate.getDay();

  return WEEK_DAYS[weekDayIndex];
};
