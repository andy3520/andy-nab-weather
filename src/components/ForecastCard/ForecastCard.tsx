import React from 'react';
import { WEATHER_STATES } from '../../constants/weather';
import { DisplayWeatherDetail, IWeatherLoc } from '../../types';
import { getRoundNum } from '../../utils/getRoundNum';
import { getWeekDay } from '../../utils/getWeekDay';

export interface IForecastCardProps {
  currentWeatherIndex: number;
  weather: IWeatherLoc;
}

const ForecastCard: React.FC<IForecastCardProps> = ({
  currentWeatherIndex,
  weather,
}) => {
  const currentWeather: DisplayWeatherDetail =
    weather?.consolidated_weather?.[currentWeatherIndex];

  if (!currentWeather) return null;

  const { title: location } = weather;
  const {
    applicable_date,
    max_temp,
    min_temp,
    the_temp,
    weather_state_abbr,
  } = currentWeather;
  const weatherState = WEATHER_STATES?.[weather_state_abbr];

  return (
    <div className="px-4 py-5 rounded-lg shadow bg-gradient-to-br sm:bg-gradient-to-r from-indigo-600 to-blue-500 sm:px-6">
      <div className="flex flex-wrap items-center justify-center sm:justify-between">
        <div className="flex-1 text-white md:pr-6">
          <h2 className="text-lg font-semibold sm:text-xl">
            {location} Weather
          </h2>
          <p className="text-base text-gray-100 sm:text-lg">
            {getWeekDay(applicable_date)} - {applicable_date}
          </p>
          <p className="font-bold uppercase text-7xl sm:text-9xl">
            {getRoundNum(the_temp)}°
          </p>
          <p className="text-xl font-semibold text-gray-100 sm:text-2xl">
            {getRoundNum(min_temp)}° &#8644; {getRoundNum(max_temp)}°
          </p>
        </div>
        <img
          src={`/static/images/weather/${weather_state_abbr}.svg`}
          title={weatherState}
          alt={`${weatherState} svg icon`}
          className="w-20 h-20 mt-6 sm:w-40 sm:h-40 sm:mt-0 animate-pulse"
        />
      </div>
    </div>
  );
};

export default ForecastCard;
