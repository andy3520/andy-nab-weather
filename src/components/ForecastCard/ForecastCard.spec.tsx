import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import ForecastCard, { EForecastCardTestId } from './ForecastCard';
import { locWeatherSample } from '../../constants/testSample';
import { getWeekDay } from '../../utils/getWeekDay';
import { getRoundNum } from '../../utils/getRoundNum';
import { WEATHER_STATES } from '../../constants/weather';

describe('ForecastCard', () => {
  const weather = locWeatherSample;

  test('should match snapshot', () => {
    const currentWeatherIndex = 0;
    const { asFragment } = render(
      <ForecastCard
        currentWeatherIndex={currentWeatherIndex}
        weather={weather}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should not render', () => {
    const { asFragment } = render(
      <ForecastCard currentWeatherIndex={0} weather={undefined} />
    );
    expect(asFragment().firstChild).toBeNull();
  });

  test.each`
    currentWeatherIndex
    ${0}
    ${1}
    ${2}
    ${3}
    ${4}
    ${5}
  `(
    'should render data dayIndex $currentWeatherIndex correctly',
    ({ currentWeatherIndex }) => {
      const expectedTitle = `${weather.title} Weather`;
      const expectCardWeather =
        weather.consolidated_weather[currentWeatherIndex];
      const { getByTestId } = render(
        <ForecastCard
          currentWeatherIndex={currentWeatherIndex}
          weather={weather}
        />
      );

      const TITLE = getByTestId(EForecastCardTestId.TITLE);
      const DATE = getByTestId(EForecastCardTestId.DATE);
      const WEEK_DAY = getByTestId(EForecastCardTestId.WEEK_DAY);
      const THE_TEMP = getByTestId(EForecastCardTestId.THE_TEMP);
      const MIN_TEMP = getByTestId(EForecastCardTestId.MIN_TEMP);
      const MAX_TEMP = getByTestId(EForecastCardTestId.MAX_TEMP);
      const IMAGE = getByTestId(EForecastCardTestId.IMAGE);

      expect(TITLE).toHaveTextContent(expectedTitle);
      expect(DATE).toHaveTextContent(expectCardWeather.applicable_date);
      expect(WEEK_DAY).toHaveTextContent(
        getWeekDay(expectCardWeather.applicable_date)
      );
      expect(THE_TEMP).toHaveTextContent(
        getRoundNum(expectCardWeather.the_temp).toString()
      );
      expect(MIN_TEMP).toHaveTextContent(
        getRoundNum(expectCardWeather.min_temp).toString()
      );
      expect(MAX_TEMP).toHaveTextContent(
        getRoundNum(expectCardWeather.max_temp).toString()
      );
      expect(IMAGE).toHaveAttribute(
        'alt',
        `${WEATHER_STATES[expectCardWeather.weather_state_abbr]} icon`
      );
    }
  );
});
