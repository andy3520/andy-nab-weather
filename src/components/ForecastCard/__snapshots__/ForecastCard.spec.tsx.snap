// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastCard should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="px-4 py-5 rounded-lg shadow bg-gradient-to-br sm:bg-gradient-to-r from-indigo-600 to-blue-500 sm:px-6"
  >
    <div
      class="flex flex-wrap items-center justify-center sm:justify-between"
    >
      <div
        class="flex-1 text-white md:pr-6"
      >
        <h2
          class="text-lg font-semibold sm:text-xl"
          data-testid="ForecastCard_TITLE"
        >
          Ho Chi Minh City Weather
        </h2>
        <p
          class="text-base text-gray-100 sm:text-lg"
        >
          <span
            data-testid="ForecastCard_WEEK_DAY"
          >
            Friday
          </span>
           - 
          <span
            data-testid="ForecastCard_DATE"
          >
            2021-01-01
          </span>
        </p>
        <p
          class="font-bold uppercase text-7xl sm:text-9xl"
        >
          <span
            data-testid="ForecastCard_THE_TEMP"
          >
            28
          </span>
          °
        </p>
        <p
          class="text-xl font-semibold text-gray-100 sm:text-2xl"
        >
          <span
            data-testid="ForecastCard_MIN_TEMP"
          >
            21
          </span>
          ° ⇄ 
          <span
            data-testid="ForecastCard_MAX_TEMP"
          >
            29
          </span>
          °
        </p>
      </div>
      <img
        alt="Heavy Cloud icon"
        class="w-20 h-20 mt-6 sm:w-40 sm:h-40 sm:mt-0 animate-pulse"
        data-testid="ForecastCard_IMAGE"
        src="/static/images/weather/hc.svg"
        title="Heavy Cloud"
      />
    </div>
  </div>
</DocumentFragment>
`;
